server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala

```diff
@@ -376,6 +376,9 @@ class InteractiveSession(
     }
     uriFuture onFailure { case e => warn("Fail to get rsc uri", e) }
 
+    // Register this class to RSCClient as a session state listener
+    client.get.registerStateListener(this)
```

Comment:
I think normally this `register` should be invoked earlier than `RSCClient` connecting to remote driver. This is a very fast non-block operation compared to network connection operations in `RSCClient`.